core/server/master/src/main/java/alluxio/master/file/LostFileDetector.java

```diff
@@ -37,34 +41,44 @@ final class LostFileDetector implements HeartbeatExecutor {
   private static final Logger LOG = LoggerFactory.getLogger(LostFileDetector.class);
   private final FileSystemMaster mFileSystemMaster;
   private final InodeTree mInodeTree;
+  private final BlockMaster mBlockMaster;
 
   /**
    * Constructs a new {@link LostFileDetector}.
    */
-  public LostFileDetector(FileSystemMaster fileSystemMaster, InodeTree inodeTree) {
+  public LostFileDetector(FileSystemMaster fileSystemMaster, BlockMaster blockMaster,
+      InodeTree inodeTree) {
     mFileSystemMaster = fileSystemMaster;
     mInodeTree = inodeTree;
+    mBlockMaster = blockMaster;
     MetricsSystem.registerCachedGaugeIfAbsent(MetricKey.MASTER_LOST_FILE_COUNT.getName(),
         () -> mFileSystemMaster.getLostFiles().size());
   }
 
   @Override
   public void heartbeat() throws InterruptedException {
-    for (long fileId : mFileSystemMaster.getLostFiles()) {
+    Iterator<Long> iter = mBlockMaster.getLostBlocksIterator();
+    while (iter.hasNext()) {
+      long blockId = iter.next();
+      // the file id is the container id of the block id
+      long containerId = BlockId.getContainerId(blockId);
+      long fileId = IdUtils.createFileId(containerId);
       // Throw if interrupted.
       if (Thread.interrupted()) {
         throw new InterruptedException("LostFileDetector interrupted.");
       }
       // update the state
       try (JournalContext journalContext = mFileSystemMaster.createJournalContext();
-          LockedInodePath inodePath =
-              mInodeTree.lockFullInodePath(fileId, LockPattern.WRITE_INODE)) {
+           LockedInodePath inodePath =
+               mInodeTree.lockFullInodePath(fileId, LockPattern.WRITE_INODE)) {
```

Comment:
Actually I think we can start with locking with the read lock because we just want to check the persistence state. If it turns out we need to update the inode, then we can release the read lock and re-acquire the WRITE_INODE lock.